staging/kos/pkg/controllers/subnet/validator_test.go

```diff
@@ -0,0 +1,506 @@
+// It does not use the testing scaffolding used by other tests and it exposes more
+// implementation wrt the other tests. Think whether we can harmonize.
+func TestSubnetValidator_lateInformer(t *testing.T) {
+	// s1 is the subnet that will be processed.
+	s1 := newSubnet1()
+
+	// s2 is s1's rival, their CIDRs overlap.
+	s2 := newSubnet1(uid(uid2), name(name2))
+
+	// Put s1 and s2 in the client (AKA the fake API server).
+	client := kosfake.NewSimpleClientset(s1, s2)
+
+	subnetsInformer := kosinformers.NewSharedInformerFactory(client, 0).Network().V1alpha1().Subnets()
+
+	// Put only s1 in the informer: this allows simulating the scenario where s1
+	// is processed with s2 not yet in the informer.
+	subnetsInformer.Informer().GetStore().Add(s1)
+
+	subnetValidator := NewValidationController(client.NetworkV1alpha1(),
+		subnetsInformer.Informer(),
+		subnetsInformer.Lister(),
+		nil,
+		workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(0, 0)),
+		0,
+		"",
+		true)
+
+	// Process s1. Ignore the error that will be returned (assuming the
+	// validator is correct) since in the current implementation successful
+	// processing of a subnet is conditional to the presence of data structures
+	// initialized when rival subnets are processed (and s2 is never processed).
+	// TODO: Ignoring the error gives away a lot of the implementation.
+	subnetValidator.processSubnet(k8stypes.NamespacedName{s1.Namespace, s1.Name})
+
+	// Retrieve s1 after processing it.
+	gotS1, err := client.NetworkV1alpha1().Subnets(s1.Namespace).Get(s1.Name, k8smetav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("error while retrieving s1: %s", err.Error())
+	}
+
+	// s1 MUST NOT be validated even if its only rival is not in the informer.
+	if gotS1.Status.Validated {
+		t.Fatalf("subnet was marked as valid even if rival exists")
+	}
+}
+
+// incRVOnUpdate assumes that `action` is a subnet update and that the subnet's
+// resource version can be parsed into an int, never call it if such conditions
+// are not met.
+func incRVOnUpdate(action k8stesting.Action) (bool, runtime.Object, error) {
+	update := action.(k8stesting.UpdateAction)
+	subnet := update.GetObject().(*netv1a1.Subnet)
+	rv, _ := strconv.Atoi(subnet.ResourceVersion)
+	rv++
+	subnet.ResourceVersion = strconv.Itoa(rv)
+	return false, nil, nil
+}
+
+// newUpdateErrorReaction can be registered to intercept updates from the fake
+// client used for tests and return a failure. It allows testing how the
+// validator reacts to updates failures.
+func newUpdateErrorReaction(numberOfErrors int) k8stesting.ReactionFunc {
+	i := 0
+	return func(action k8stesting.Action) (bool, runtime.Object, error) {
+		if i < numberOfErrors {
+			i++
+			return true, nil, errors.New("test error")
+		}
+		return false, nil, nil
+	}
+}
+
+//****************************************************************************
+// Follow functional options and util functions to create test subnets with  //
```

Comment:
"The following functional options and util functions create ..."

Reply:
Fixed in 7ce6161 .